elliottlib/util.py

```diff
@@ -238,6 +238,33 @@ def parallel_results_with_progress(inputs, func, file=None):
     return results
 
 
+def get_target_release(bugs):
```

Comment:
This function gets a list of bugs attached to an advisory, it then validates that the target release of all those bugs is the one single exact value example `4.7.0`, cannot be `4.7.z`. I assumed that was the right logic. Is that assumption correct?

Reply:
I think it's a bit more complicated than that I'm afraid.

Each flaw bug can have many trackers:
```
flaw1 -> tracker1
      -> tracker2
      -> tracker3
      -> tracker4
      -> tracker5           
```

Each tracker can have any value for it's target release:
```
flaw1 -> tracker1: 4.7.0
      -> tracker2: 4.7.0
      -> tracker3: 4.8.0
      -> tracker4: 4.7.z
      -> tracker5: 4.6.0
```

To work out if it's a first fix or not we need to look at the _component_:
```
flaw1 -> podman:  4.7.0
      -> cri-o:   4.7.0
      -> skopeo:  4.8.0
      -> skopeo:  4.7.z
      -> buildah: 4.6.0
```
For 3/4 components above, they only have one bug so that bug represents the first time that component will be fixed, i.e. it will be a first fix:

```
flaw1 -> podman:  4.7.0 (first fix)
      -> cri-o:   4.7.0 (first fix)
      -> skopeo:  4.8.0
      -> skopeo:  4.7.z
      -> buildah: 4.6.0 (first fix)
```
Only for skopeo do we need extra checks, because there is more than one tracker bug for skopeo. _If_ the 4.7.z skopeo bug has been fixed, only then will the 4.8.0 skopeo _not_ be a first fix (because it has been fixed already in z-stream). If the 4.7.z skopeo bug has not been fixed, then the 4.8.0 skopeo bug will be the first fix.

NB: z-stream bugs area always, always first fixes. We only need these first fix checks for .0 bugs, e.g. 4.7.0.

Reply:
TL;DR i think is that aside from

> We only need these first fix checks for .0 bugs, e.g. 4.7.0.

... 4.y.0 and 4.y.z are equivalent.

Reply:
that said, this was just about making sure all the bugs on the advisory are for the same target, not the trackers for a flaw bug.

yes, there should be a unique answer - 4.y.0 bugs should change to 4.y.z bugs if they don't make GA.

Reply:
So it looks to me that we're trying to answer the question **Which (ocp) release is considered the "first-fix" release for the component X for the cve flaw Y?** <- let me know if this is accurate

This means that our first fix logic needs to be enhanced with what you describe. Right now we're checking if any tracker bug for a flaw bug has been fixed (or is being fixed), then it's not a first-fix, otherwise it is. The component checks is missing.

Reply:
I think our definition is stricter, it just takes flaw and tracker release versions. So right now we'll be missing attaching flaw bugs

Reply:
> So it looks to me that we're trying to answer the question Which (ocp) release is considered the "first-fix" release for the component X for the cve flaw Y? <- let me know if this is accurate

There can, and often are, many first fixes. I.e. z-stream fixes are always first fixes for that given stream. It's only possible for a .0 fix to be a non-first fix for the entire product, if it's already been fixed in an earlier z-stream. In that case, the .0 release would not be a security update for the customer.

.0 fixes are only fist fixes, if it has not yet been fixed in an earlier z-stream.

Reply:
thanks sam!